src/derived.rs

```diff
@@ -46,6 +44,18 @@ where
 {
     map: RwLock<FxHashMap<Q::Key, QueryState<DB, Q>>>,
     policy: PhantomData<MP>,
+
+    /// This cond var is used when one thread is waiting on another to
+    /// produce some specific key. In that case, the thread producing
+    /// the key will signal the cond-var. The threads awaiting the key
+    /// will check in `map` to see if their key is present and (if
+    /// not) await the cond-var.
+    signal_cond_var: Condvar,
```

Comment:
This is a per-table cond-var, so we potentially wake to many threads, right?

Curious, would it be easier/more efficient to implement per-key notification? That is, could we add some struct which a thread can wait on to the `InProgress` variant?

For example, I think we can use crossbeam-channels for this sort of notification: we store an `Receiver<Void>` in the `InProgress` struct:

```
InProgress {
  id: RuntimeId,
  ready: Receiver<Void>,
}
```

When we store the in-progress placeholder, we create a Sender/Receiver pair and make sure to drop the Sender only after we've overwritten the placeholder with value.

When we read a placeholder, we do `ready.recv()` to block until the results are ready. 

That way, all subtle notification and waiting logic is delegated to crossbeam channel. 


In other words, these signals, mutexes and atomic variables add an entirely new dimension of subtleness, and it would be cool to abstract them away with something more high-level :)

Reply:
However, I don't think that the setup you described will work: you would need a channel per receiver, from what I can tell. 

What I was considering was having a `Mutex<Vec<Receiver<T>>>` in the `InProgress`, initially empty (so cheap to construct). Receivers would grab the mutex and insert themselves, then release all locks and wait for a message. 

After completing, senders would unwrap the mutex, extract the vector, and send clones to the value. 

This also handles the "no memoized value" case.

Reply:
>you would need a channel per receiver, from what I can tell.

If we want to use `Receiver<T>`, that is, to send an actual value, then yes, we need several channels, because they don't have broadcast. 

However, I was suggesting using channels *only* for synchronization, using `!` as a message type. That way, you'd have n receivers, one sender, and, when the sender is dropped, all receivers are notified, but have to look up the actual value in the map themselves. 

,
Agree though that this approach would have to allocate for every re computation of a query which, on one hand, seems sub optimal, but on the other hand, we probably will allocate an Arc for the final result anyway.

Reply:
Hm, I think we can tweak the channel approach so that we allocate the channel only when the blocking actually happens? That is, we can change `InProgress` to something like

```
InProgress {
    id: RuntimeId,
    ready: Option<(Sender<!>, Receiver<!>)>,
}
```

The first thread to see `InProgress` then allocates the channels, subsequent threads then clone the receiver, and the notification happens naturally when the `InProgress` value is dropped. Not sure that these additional logic makes this solution more complex than the current one.

Reply:
OK, had a different (but similar) idea that I will experiment with later